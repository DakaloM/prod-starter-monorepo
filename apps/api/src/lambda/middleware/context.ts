import { Handler, APIGatewayEventV2 } from '@imax/serverlesskit';

import middy, { MiddlewareObj } from '@middy/core';
import knex from 'knex';
import { AuthContext } from '~/auth';
import { DatabaseCache } from '~/cache/cache';
import { Context } from '~/context';
import { knexfile } from '~/knex';
import { mailer } from '~/mailer';
import { storageClient } from '~/storage';

const db = knex(knexfile);
export function addContext<TEvent extends ContextEvent>(): MiddlewareObj<TEvent> {
  return {
    before: async (request) => {
      const header = request.event.headers['authorization'] || '';
      const token = header.replace('Bearer ', '');

      const ctx = Context.init({
        db,
        mailer: mailer,
        cache: new DatabaseCache(db),
        storageClient,
      });

      const auth = await AuthContext.init(token, ctx);

      ctx.auth = auth;

      request.event.ctx = ctx;
    },
  };
}

export function withContext<TBody, TRes>(handler: Handler<TBody, TRes>) {
  return middy(handler).use(addContext());
}

export type ContextEvent = APIGatewayEventV2 & {
  ctx: Context;
};
