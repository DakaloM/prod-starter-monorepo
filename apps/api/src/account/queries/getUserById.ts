import { NotFoundError } from '@imax/serverkit';

import { User } from '~/account/user';
import { Context } from '~/context';

export async function getUserById(id: string, ctx: Context) {
  const user = await User.query(ctx.db).findById(id);

  if (!user) {
    throw new NotFoundError({
      message: 'User not found',
    });
  }

  return user;
}

