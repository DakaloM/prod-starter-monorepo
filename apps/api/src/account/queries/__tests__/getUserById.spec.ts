import crypto from 'node:crypto';
import { Context } from '~/context';
import { createContext, destroyContext } from '~/test/context';
import { Environment, createEnv, destroyEnv } from '~/test/environment';
import { Factory, createFactory } from '~/test/factory';

import { getUserById } from '../getUserById';

describe('account/queries/getUserById', () => {
  let env: Environment;
  let ctx: Context;
  let factory: Factory;
  beforeAll(async () => {
    env = await createEnv();
  });

  afterAll(async () => {
    await destroyEnv(env);
  });

  beforeEach(async () => {
    ctx = await createContext(env);
    factory = createFactory(ctx);
  });

  afterEach(async () => {
    await destroyContext(ctx);
  });

  it('should return a user', async () => {
    const user = await factory.insert('user');

    const result = await getUserById(user.id, ctx);

    expect(result.email).toEqual(user.email);
  });

  it('should throw error when user does not exist', async () => {
    const id = crypto.randomUUID();

    await expect(getUserById(id, ctx)).rejects.toMatchSnapshot();
  });
});
