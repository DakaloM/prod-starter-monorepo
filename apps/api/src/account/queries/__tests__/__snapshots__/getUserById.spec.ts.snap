// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`account/queries/getUserById > should throw error when user does not exist 1`] = `[Error: User not found]`;
